src/MOCKernel.cpp

```diff
@@ -0,0 +1,156 @@
+ */
+void MOCKernel::newTrack(Track* track) {
+  _count = 0;
+}
+
+
+/**
+ * @brief Prepares a VolumeKernel for a new Track
+ * @details Resets the segment count and updates the weight for the new Track
+ * @param track The new Track the MOCKernel prepares to handle
+ */
+void VolumeKernel::newTrack(Track* track) {
+
+  /* Compute the Track cross-sectional area */
+  int azim_index = track->getAzimAngleIndex();
+  _weight = _quadrature->getAzimSpacing(azim_index)
+      * _quadrature->getAzimWeight(azim_index);
+
+  /* Reset the count */
+  _count = 0;
+}
+
+
+/**
+ * @brief Destructor for MOCKernel
+ */
+MOCKernel::~MOCKernel() {};
+
+
+/*
+ * @brief Reads and returns the current count
+ * @details MOC kernels count how many times they are accessed. This value
+ *          returns the value of the counter (number of execute accesses)
+ *          since kernel creation or last call to newTrack.
+ * @return the number of segments traversed
+ */
+int MOCKernel::getCount() {
+  return _count;
+}
+
+
+/* @brief Resets the maximum optcal path length for a segment
+ * @details MOC kernels ensure that there are no segments with an optical path
+ *          length greater than the maximum optical path length by splitting
+ *          then when they get too large.
+ * @param max_tau the maximum optical path length for a segment
+ */
+void MOCKernel::setMaxOpticalLength(FP_PRECISION max_tau) {
+  _max_tau = max_tau;
+}
+
+
+/*
+ * @brief Adds segment contribution to the FSR volume
+ * @details The VolumeKernel execute function adds the product of the
+ *          track length and track weight to the buffer array at index
+ *          id, referring to the array of FSR volumes.
+ * @param length segment length
+ * @param mat Material associated with the segment
+ * @param id the FSR ID of the FSR associated with the segment
+ */
+void VolumeKernel::execute(FP_PRECISION length, Material* mat, int id,
+                           int cmfd_surface_fwd, int cmfd_surface_bwd) {
+
+  /* Set omp lock for FSRs */
+  omp_set_lock(&_FSR_locks[id]);
+
+  /* Add value to buffer */
+  _FSR_volumes[id] += _weight * length;
+
+  /* Unset lock */
+  omp_unset_lock(&_FSR_locks[id]);
+
+  /* Increment count */
+  _count++;
+}
+
+
+/*
+ * @brief Increments the counter for the number of segments on the track
+ * @details The CounterKernel execute function counts the number of segments
+ *          in a track by incrementing the counter variable upon execution. Due
+ *          to restrictions on maximum optical path length, the counter may be
+ *          incremented by more than one to account for splitting of the
+ *          segment into segments of allowed optical path length.
+ * @param length segment length
+ * @param mat Material associated with the segment
+ * @param id the FSR ID of the FSR associated with the segment
+ */
+void CounterKernel::execute(FP_PRECISION length, Material* mat, int id,
+    int cmfd_surface_fwd, int cmfd_surface_bwd) {
```

Comment:
it looks like this continuation line for the method signature isn't tabbed over inline with the line above